File: sdk/c/src/providers/tempmailpro.c
/**
 * TempMail Pro -- https://tempmailpro.us
 */

#include "tempmail_internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define TEMPMAILPRO_BASE "https://tempmailpro.us/api/v1"

static const char *tempmailpro_headers[] = {
    "Accept: application/json",
    "User-Agent: Mozilla/5.0",
    NULL
};

static const char *tempmailpro_post_headers[] = {
    "Accept: application/json",
    "Content-Type: application/json",
    "User-Agent: Mozilla/5.0",
    NULL
};

static char *tempmailpro_urlencode(const char *s) {
    if (!s) return tm_strdup("");
    size_t len = strlen(s);
    char *out = (char *)malloc(len * 3 + 1);
    if (!out) return NULL;
    char *p = out;
    static const char hex[] = "0123456789ABCDEF";
    for (size_t i = 0; i < len; i++) {
        unsigned char c = (unsigned char)s[i];
        if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
            (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~') {
            *p++ = (char)c;
        } else {
            *p++ = '%';
            *p++ = hex[c >> 4];
            *p++ = hex[c & 15];
        }
    }
    *p = '\0';
    return out;
}

static cJSON *tempmailpro_json(tm_http_method_t method, const char *url, const char *body) {
    const char **headers = method == TM_HTTP_POST ? tempmailpro_post_headers : tempmailpro_headers;
    tm_http_response_t *resp = tm_http_request(method, url, headers, body, 15);
    if (!resp || resp->status < 200 || resp->status >= 300) {
        tm_http_response_free(resp);
        return NULL;
    }
    cJSON *root = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    return root;
}

tm_email_info_t *tm_provider_tempmailpro_generate(void) {
    cJSON *root = tempmailpro_json(TM_HTTP_POST, TEMPMAILPRO_BASE "/mailbox/create", "{}");
    if (!root) return NULL;
    cJSON *data = cJSON_GetObjectItemCaseSensitive(root, "data");
    const char *address = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(data, "address"), "");
    const char *token = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(data, "token"), "");
    if (!address[0] || !strchr(address, '@') || !token[0]) {
        cJSON_Delete(root);
        return NULL;
    }

    tm_email_info_t *info = tm_email_info_new();
    if (!info) {
        cJSON_Delete(root);
        return NULL;
    }
    info->channel = CHANNEL_TEMPMAILPRO;
    info->email = tm_strdup(address);
    info->token = tm_strdup(token);
    cJSON *expires_at = cJSON_GetObjectItemCaseSensitive(data, "expires_at");
    if (cJSON_IsNumber(expires_at)) info->expires_at = (long long)expires_at->valuedouble;
    cJSON *created_at = cJSON_GetObjectItemCaseSensitive(data, "created_at");
    if (cJSON_IsString(created_at) && created_at->valuestring) {
        info->created_at = tm_strdup(created_at->valuestring);
    } else if (cJSON_IsNumber(created_at)) {
        char buf[64];
        snprintf(buf, sizeof(buf), "%.0f", created_at->valuedouble);
        info->created_at = tm_strdup(buf);
    }
    cJSON_Delete(root);
    return info;
}

static cJSON *tempmailpro_fetch_detail(const char *token, const char *message_id) {
    char *token_enc = tempmailpro_urlencode(token);
    char *id_enc = tempmailpro_urlencode(message_id);
    if (!token_enc || !id_enc) {
        free(token_enc);
        free(id_enc);
        return NULL;
    }
    size_t cap = strlen(TEMPMAILPRO_BASE) + strlen(token_enc) + strlen(id_enc) + 32;
    char *url = (char *)malloc(cap);
    if (!url) {
        free(token_enc);
        free(id_enc);
        return NULL;
    }
    snprintf(url, cap, "%s/mailbox/%s/emails/%s", TEMPMAILPRO_BASE, token_enc, id_enc);
    free(token_enc);
    free(id_enc);

    cJSON *root = tempmailpro_json(TM_HTTP_GET, url, NULL);
    free(url);
    if (!root) return NULL;
    cJSON *data = cJSON_GetObjectItemCaseSensitive(root, "data");
    cJSON *dup = cJSON_IsObject(data) ? cJSON_Duplicate(data, 1) : NULL;
    cJSON_Delete(root);
    return dup;
}

static cJSON *tempmailpro_message_raw(cJSON *msg, const char *recipient) {
    cJSON *raw = cJSON_CreateObject();
    if (!raw) return NULL;
    const cJSON *id = cJSON_GetObjectItemCaseSensitive(msg, "id");
    if (!id) id = cJSON_GetObjectItemCaseSensitive(msg, "message_id");
    const cJSON *from = cJSON_GetObjectItemCaseSensitive(msg, "from_address");
    if (!from) from = cJSON_GetObjectItemCaseSensitive(msg, "from_name");
    cJSON_AddStringToObject(raw, "id", TM_JSON_STR(id, ""));
    cJSON_AddStringToObject(raw, "from", TM_JSON_STR(from, ""));
    cJSON_AddStringToObject(raw, "to", recipient);
    cJSON_AddStringToObject(raw, "subject", TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(msg, "subject"), ""));
    cJSON_AddStringToObject(raw, "text", TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(msg, "body_text"), ""));
    cJSON_AddStringToObject(raw, "html", TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(msg, "body_html"), ""));
    cJSON *received_at = cJSON_GetObjectItemCaseSensitive(msg, "received_at");
    if (cJSON_IsNumber(received_at)) cJSON_AddNumberToObject(raw, "date", received_at->valuedouble);
    else cJSON_AddStringToObject(raw, "date", TM_JSON_STR(received_at, ""));
    cJSON *attachments = cJSON_GetObjectItemCaseSensitive(msg, "attachments");
    if (attachments) cJSON_AddItemToObject(raw, "attachments", cJSON_Duplicate(attachments, 1));
    return raw;
}

tm_email_t *tm_provider_tempmailpro_get_emails(const char *token, const char *email, int *count) {
    *count = -1;
    if (!token || !token[0] || !email || !email[0]) return NULL;
    char *token_enc = tempmailpro_urlencode(token);
    if (!token_enc) return NULL;
    size_t cap = strlen(TEMPMAILPRO_BASE) + strlen(token_enc) + 24;
    char *url = (char *)malloc(cap);
    if (!url) {
        free(token_enc);
        return NULL;
    }
    snprintf(url, cap, "%s/mailbox/%s/emails", TEMPMAILPRO_BASE, token_enc);
    free(token_enc);

    cJSON *root = tempmailpro_json(TM_HTTP_GET, url, NULL);
    free(url);
    if (!root) return NULL;
    cJSON *data = cJSON_GetObjectItemCaseSensitive(root, "data");
    if (!cJSON_IsArray(data)) {
        *count = 0;
        cJSON_Delete(root);
        return NULL;
    }
    int n = cJSON_GetArraySize(data);
    *count = n;
    if (n == 0) {
        cJSON_Delete(root);
        return NULL;
    }

    tm_email_t *emails = tm_emails_new(n);
    if (!emails) {
        *count = -1;
        cJSON_Delete(root);
        return NULL;
    }
    for (int i = 0; i < n; i++) {
        cJSON *item = cJSON_GetArrayItem(data, i);
        const char *message_id = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(item, "id"), "");
        cJSON *detail = message_id[0] ? tempmailpro_fetch_detail(token, message_id) : NULL;
        cJSON *raw = tempmailpro_message_raw(detail ? detail : item, email);
        emails[i] = tm_normalize_email(raw ? raw : (detail ? detail : item), email);
        if (raw) cJSON_Delete(raw);
        if (detail) cJSON_Delete(detail);
    }
    cJSON_Delete(root);
    return emails;
}

File: sdk/rust/src/providers/tempmailpro.rs
/*!
 * TempMail Pro - https://tempmailpro.us
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::{json, Value};

const API_BASE: &str = "https://tempmailpro.us/api/v1";

fn flatten(raw: &Value, recipient: &str) -> Value {
    json!({
        "id": raw.get("id").cloned().or_else(|| raw.get("message_id").cloned()).unwrap_or(Value::String(String::new())),
        "from": raw.get("from_address").cloned().or_else(|| raw.get("from_name").cloned()).unwrap_or(Value::String(String::new())),
        "to": recipient,
        "subject": raw.get("subject").cloned().unwrap_or(Value::String(String::new())),
        "text": raw.get("body_text").cloned().unwrap_or(Value::String(String::new())),
        "html": raw.get("body_html").cloned().unwrap_or(Value::String(String::new())),
        "date": raw.get("received_at").cloned().unwrap_or(Value::Null),
        "attachments": raw.get("attachments").cloned().unwrap_or(Value::Array(Vec::new())),
    })
}

pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let resp = http_client()
            .post(format!("{}/mailbox/create", API_BASE))
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .header("User-Agent", "Mozilla/5.0")
            .json(&json!({}))
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("tempmailpro http {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        let box_data = data.get("data").unwrap_or(&Value::Null);
        let email = box_data
            .get("address")
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        let token = box_data
            .get("token")
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        if email.is_empty() || !email.contains('@') || token.is_empty() {
            return Err("tempmailpro: invalid mailbox response".into());
        }
        Ok(EmailInfo {
            channel: Channel::Tempmailpro,
            email,
            token: Some(token),
            expires_at: box_data.get("expires_at").and_then(|x| x.as_i64()),
            created_at: box_data.get("created_at").map(|x| {
                x.as_str()
                    .map(|s| s.to_string())
                    .unwrap_or_else(|| x.to_string())
            }),
        })
    })
}

fn fetch_detail(token: &str, message_id: &str) -> Result<Value, String> {
    block_on(async {
        let resp = http_client()
            .get(format!(
                "{}/mailbox/{}/emails/{}",
                API_BASE,
                urlencoding::encode(token),
                urlencoding::encode(message_id)
            ))
            .header("Accept", "application/json")
            .header("User-Agent", "Mozilla/5.0")
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("tempmailpro message {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        Ok(data.get("data").cloned().unwrap_or(Value::Null))
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let mailbox_token = token.trim();
    let address = email.trim();
    if mailbox_token.is_empty() {
        return Err("tempmailpro: empty token".into());
    }
    if address.is_empty() {
        return Err("tempmailpro: empty email".into());
    }

    block_on(async {
        let resp = http_client()
            .get(format!(
                "{}/mailbox/{}/emails",
                API_BASE,
                urlencoding::encode(mailbox_token)
            ))
            .header("Accept", "application/json")
            .header("User-Agent", "Mozilla/5.0")
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("tempmailpro messages {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        let rows = data
            .get("data")
            .and_then(|x| x.as_array())
            .cloned()
            .unwrap_or_default();
        let mut out = Vec::new();
        for row in rows {
            let raw = row
                .get("id")
                .map(|x| x.to_string())
                .map(|id| id.trim_matches('"').to_string())
                .filter(|id| !id.is_empty())
                .and_then(|id| fetch_detail(mailbox_token, &id).ok())
                .filter(|x| !x.is_null())
                .unwrap_or(row);
            out.push(normalize_email(&flatten(&raw, address), address));
        }
        Ok(out)
    })
}
